client/pruning/main.go

```diff
@@ -0,0 +1,118 @@
+		custom: allow pruning options to be manually specified through 'pruning-keep-recent'.
+
+		besides pruning options, database home directory and database backend type should also be specified via flags
+		'--home' and '--app-db-backend'.
+		valid app-db-backend type includes 'goleveldb', 'cleveldb', 'rocksdb', 'boltdb', and 'badgerdb'.
+		`,
+		Example: "prune --home './' --app-db-backend 'goleveldb' --pruning 'custom' --pruning-keep-recent 100",
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			vp := viper.New()
+
+			// Bind flags to the Context's Viper so we can get pruning options.
+			if err := vp.BindPFlags(cmd.Flags()); err != nil {
+				return err
+			}
+			pruningOptions, err := server.GetPruningOptionsFromFlags(vp)
+			if err != nil {
+				return err
+			}
+			fmt.Printf("get pruning options from command flags, strategy: %v, keep-recent: %v\n",
+				pruningOptions.Strategy,
+				pruningOptions.KeepRecent,
+			)
+
+			home := vp.GetString(flags.FlagHome)
+			db, err := openDB(home, server.GetAppDBBackend(vp))
+			if err != nil {
+				return err
+			}
+
+			logger := log.NewTMLogger(log.NewSyncWriter(os.Stdout))
+			app := appCreator(logger, db, nil, vp)
+			cms := app.CommitMultiStore()
+
+			rootMultiStore, ok := cms.(*rootmulti.Store)
+			if !ok {
+				return fmt.Errorf("currently only support the pruning of rootmulti.Store type")
+			}
+			latestHeight := rootmulti.GetLatestVersion(db)
+			// valid heights should be greater than 0.
+			if latestHeight <= 0 {
+				return fmt.Errorf("the database has no valid heights to prune, the latest height: %v", latestHeight)
+			}
+
+			var pruningHeights []int64
+			for height := int64(1); height < latestHeight; height++ {
+				if height < latestHeight-int64(pruningOptions.KeepRecent) {
+					pruningHeights = append(pruningHeights, height)
+				}
+			}
+			if len(pruningHeights) == 0 {
+				fmt.Printf("no heights to prune\n")
+				return nil
+			}
+			fmt.Printf(
+				"pruning heights start from %v, end at %v\n",
+				pruningHeights[0],
+				pruningHeights[len(pruningHeights)-1],
+			)
+
+			err = rootMultiStore.PruneStores(false, pruningHeights)
+			if err != nil {
+				return err
+			}
+			fmt.Printf("successfully pruned the application root multi stores\n")
+			return nil
+		},
+	}
+
+	cmd.Flags().String(flags.FlagHome, "", "The database home directory")
+	cmd.Flags().String(FlagAppDBBackend, "", "The type of database for application and snapshots databases")
+	cmd.Flags().String(server.FlagPruning, pruningtypes.PruningOptionDefault, "Pruning strategy (default|nothing|everything|custom)")
+	cmd.Flags().Uint64(server.FlagPruningKeepRecent, 0, "Number of recent heights to keep on disk (ignored if pruning is not 'custom')")
+	cmd.Flags().Uint64(server.FlagPruningInterval, 0,
+		`Height interval at which pruned heights are removed from disk (ignored if pruning is not 'custom'), 
+		this is not used by this command but kept for compability with the complete pruning options`)
```

Comment:
I misunderstood the meaning of `pruning-interval` at first time.
actually `pruning-interval` is only used to decide at which height the pruning action would happen.
but actually for every height whether it is to-be-pruned only depends on the `keep-recent` and the `snapshot-options`, so almost every height older than `keep-recent` will be moved to `pruneHeights` after the snapshot completes.
@p0mvn Could you help confirm the pruning logic here?

Reply:
correct